tsa-core/src/main/kotlin/org/usvm/machine/interpreter/TvmArtificialInstInterpreter.kt

```diff
@@ -102,10 +124,88 @@ class TvmArtificialInstInterpreter(
+            // sending bounced messages is only considered when the message handling ended with an exception
+            // if we stop on the first error, the potential bouncing won't be considered
+            scope.doWithState {
+                newStmt(TsaArtificialExitInst(stmt.computePhaseResult, lastStmt.location))
+            }
+            return
+        }
+        scope.calcOnState {
+            with(ctx) {
+                if (result.isExceptional()) {
+                    val (inputMessage, sender, _) = receivedMessage as? ReceivedMessage.MessageFromOtherContract
+                        ?: run {
+                            newStmt(TsaArtificialExitInst(stmt.computePhaseResult, lastStmt.location))
+                            return@calcOnState
+                        }
+                    // if is bounceable, bounce back to sender
+                    val fullMsgData = cellDataFieldManager.readCellData(scope, inputMessage.fullMsgCell)
+                        ?: return@calcOnState
+                    val isBounceable = mkBvAndExpr(
+                        fullMsgData,
+                        mkBvShiftLeftExpr(oneCellValue, 1020.toCellSort())
+                    )
+
+                    val bouncedMessageCell = constructBouncedMessage(fullMsgData, scope)
+                    scope.fork(
+                        isBounceable.neq(zeroCellValue), falseStateIsExceptional = true,
+                        blockOnTrueState = {
+                            messageQueue =
+                                messageQueue.add(sender to inputMessage.copy(fullMsgCell = bouncedMessageCell))
+                            newStmt(TsaArtificialExitInst(stmt.computePhaseResult, lastStmt.location))
+                        },
+                        blockOnFalseState = {
+                            newStmt(TsaArtificialExitInst(stmt.computePhaseResult, lastStmt.location))
+
+                        }
+                    )
+                } else {
+                    scope.doWithState {
+                        newStmt(TsaArtificialExitInst(stmt.computePhaseResult, lastStmt.location))
+                    }
+                }
+            }
+        }
+    }
+
+    private fun TvmContext.constructBouncedMessage(
+        @Suppress("Unused") // tests on bounced message structure are required
+        fullMsg: UExpr<TvmContext.TvmCellDataSort>,
+        scope: TvmStepScopeManager
+    ): UHeapRef {
+        val msgCell = scope.calcOnState {
+            val content = RecvInternalInput.MessageContent(
+                flags = 0b0001.toBv257(),
+                srcAddressSlice = allocSliceFromData(0b00.toBv(2u)),
+                dstAddressSlice = allocSliceFromData(0b00.toBv(2u)),
+                msgValue = zeroValue,
+                ihrFee = zeroValue,
+                fwdFee = zeroValue,
+                createdLt = zeroValue,
+                createdAt = zeroValue,
+                bodyDataSlice = allocSliceFromData(0xFFFFFFFF.toBv(32u))
```

Comment:
Does it ignore the old message body? It should take the first 256 bits instead

Reply:
outdated